File: src/onboardingStore.test.ts
// The first-run store + policy: Skip is "not now", "Don't show me this again" silences everything, the
// seen version persists, each product + user is kept apart, and broken storage never breaks the first-run.

import { test } from 'node:test'
import assert from 'node:assert/strict'
import {
  createLocalOnboardingStore,
  DEFAULT_RESURFACE_MS,
  onboardingStorageKey,
  parseOnboardingState,
  shouldAutoOpenOnboarding,
  shouldAutoStartTour,
  type OnboardingStorage,
} from './onboardingStore.js'
import type { OnboardingStep } from './onboardingTypes.js'

const memStorage = (): OnboardingStorage & { data: Map<string, string> } => {
  const data = new Map<string, string>()
  return {
    data,
    getItem: (k) => data.get(k) ?? null,
    setItem: (k, val) => void data.set(k, val),
    removeItem: (k) => void data.delete(k),
  }
}
const steps: OnboardingStep[] = [
  { id: 'profile', title: 'Add your profile' },
  { id: 'first-report', title: 'Open your first report' },
]
const T0 = new Date('2026-09-24T12:00:00Z')

test('a fresh user gets the welcome pane', async () => {
  const store = createLocalOnboardingStore({ product: 'toudai', storage: memStorage() })
  assert.equal(shouldAutoOpenOnboarding(await store.load(), steps, T0), true)
})

test('completeStep persists and is idempotent; a finished checklist stops auto-open', async () => {
  const storage = memStorage()
  const store = createLocalOnboardingStore({ product: 'toudai', storage })
  await store.completeStep('profile')
  await store.completeStep('profile')
  const s = await store.completeStep('first-report')
  assert.deepEqual(s.completed, ['profile', 'first-report'])
  assert.equal(shouldAutoOpenOnboarding(s, steps, T0), false)
  // A new store over the same storage (the next visit) reads it back.
  const again = await createLocalOnboardingStore({ product: 'toudai', storage }).load()
  assert.deepEqual(again.completed, ['profile', 'first-report'])
})

test('Skip is "not now": closed during the quiet window, back after it while unfinished', async () => {
  const store = createLocalOnboardingStore({ product: 'toudai', storage: memStorage(), now: () => T0 })
  const s = await store.skip()
  assert.equal(s.skippedAt, T0.toISOString())
  assert.equal(shouldAutoOpenOnboarding(s, steps, new Date(T0.getTime() + 60_000)), false)
  assert.equal(shouldAutoOpenOnboarding(s, steps, new Date(T0.getTime() + DEFAULT_RESURFACE_MS + 1)), true)
  assert.equal(
    shouldAutoOpenOnboarding(s, steps, new Date(T0.getTime() + DEFAULT_RESURFACE_MS * 10), { resurfaceAfterMs: Infinity }),
    false,
  )
})

test('"Don\'t show me this again" silences the pane AND every tour, and can be undone', async () => {
  const store = createLocalOnboardingStore({ product: 'toudai', storage: memStorage() })
  let s = await store.setSuppressed(true)
  assert.equal(shouldAutoOpenOnboarding(s, steps, T0), false)
  assert.equal(shouldAutoStartTour(s, 'intro'), false)
  s = await store.setSuppressed(false)
  assert.equal(shouldAutoOpenOnboarding(s, steps, T0), true)
  assert.equal(shouldAutoStartTour(s, 'intro'), true)
})

test('tours are one-shot: finished or skipped never auto-start again; other tours still can', async () => {
  const store = createLocalOnboardingStore({ product: 'toudai', storage: memStorage() })
  let s = await store.finishTour('intro', 'skipped')
  assert.equal(shouldAutoStartTour(s, 'intro'), false)
  assert.equal(shouldAutoStartTour(s, 'reports'), true)
  s = await store.finishTour('reports', 'completed')
  assert.deepEqual(s.tours, { intro: 'skipped', reports: 'completed' })
})

test('the seen version persists across visits', async () => {
  const storage = memStorage()
  await createLocalOnboardingStore({ product: 'toudai', storage }).markVersionSeen('0.5.1865')
  assert.equal((await createLocalOnboardingStore({ product: 'toudai', storage }).load()).lastSeenVersion, '0.5.1865')
})

test('each product and each user keeps separate choices', async () => {
  const storage = memStorage()
  await createLocalOnboardingStore({ product: 'toudai', userKey: 'u1', storage }).setSuppressed(true)
  assert.equal((await createLocalOnboardingStore({ product: 'toudai', userKey: 'u2', storage }).load()).suppressed, false)
  assert.equal((await createLocalOnboardingStore({ product: 'fudemoji', userKey: 'u1', storage }).load()).suppressed, false)
  assert.ok(storage.data.has(onboardingStorageKey('toudai', 'u1')))
  assert.equal(onboardingStorageKey('ike'), 'bp:beacon:onboarding:ike:anon')
})

test('reset clears the record', async () => {
  const storage = memStorage()
  const store = createLocalOnboardingStore({ product: 'toudai', storage })
  await store.setSuppressed(true)
  const s = await store.reset()
  assert.equal(s.suppressed, false)
  assert.equal(storage.data.size, 0)
})

test('storage that throws never breaks the first-run (memory fallback for the visit)', async () => {
  const broken: OnboardingStorage = {
    getItem: () => {
      throw new Error('blocked')
    },
    setItem: () => {
      throw new Error('quota')
    },
    removeItem: () => {
      throw new Error('blocked')
    },
  }
  const store = createLocalOnboardingStore({ product: 'toudai', storage: broken })
  assert.equal((await store.load()).suppressed, false)
  assert.equal((await store.setSuppressed(true)).suppressed, true)
  assert.equal((await store.load()).suppressed, true)
  assert.equal((await store.reset()).suppressed, false)
})

test('no storage at all (SSR / node) still works in memory', async () => {
  const store = createLocalOnboardingStore({ product: 'toudai' })
  assert.deepEqual((await store.completeStep('profile')).completed, ['profile'])
})

test('a corrupt or foreign record parses field by field to safe defaults', () => {
  assert.deepEqual(parseOnboardingState('not json'), { completed: [], suppressed: false, tours: {} })
  const s = parseOnboardingState(
    JSON.stringify({ completed: ['a', 3], suppressed: 'yes', tours: { t: 'bogus', u: 'completed' }, lastSeenVersion: 1 }),
  )
  assert.deepEqual(s.completed, ['a'])
  assert.equal(s.suppressed, false)
  assert.deepEqual(s.tours, { u: 'completed' })
  assert.equal(s.lastSeenVersion, undefined)
})

File: kmp/src/commonTest/kotlin/com/bellsandpixels/beacon/onboarding/OnboardingStoreTest.kt
package com.bellsandpixels.beacon.onboarding

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

// In-memory OnboardingStorage, standing in for SharedPreferences / NSUserDefaults in commonTest.
private class MemoryStorage : OnboardingStorage {
    val map = mutableMapOf<String, String>()
    override fun getItem(key: String): String? = map[key]
    override fun setItem(key: String, value: String) { map[key] = value }
    override fun removeItem(key: String) { map.remove(key) }
}

class OnboardingStoreTest {

    private fun store(
        storage: MemoryStorage = MemoryStorage(),
        product: String = "fudemoji",
        userKey: String? = null,
        now: Long = 1_000_000L,
    ) = OnboardingStore(product, userKey, storage, nowMs = { now })

    // ---- storage key ----------------------------------------------------------------------------

    @Test
    fun storage_key_is_namespaced_by_product_and_user() {
        assertEquals("bp:beacon:onboarding:fudemoji:anon", onboardingStorageKey("fudemoji", null))
        assertEquals("bp:beacon:onboarding:fudemoji:u42", onboardingStorageKey("fudemoji", "u42"))
    }

    // ---- completeStep ---------------------------------------------------------------------------

    @Test
    fun completeStep_is_idempotent_and_persists() {
        val storage = MemoryStorage()
        val s = store(storage)
        s.completeStep("write-first-card")
        s.completeStep("write-first-card") // idempotent: no duplicate
        s.completeStep("import-deck")
        assertEquals(listOf("write-first-card", "import-deck"), s.load().completed)

        // Persisted: a fresh store over the same storage reads the same completed set.
        val reloaded = store(storage).load()
        assertEquals(listOf("write-first-card", "import-deck"), reloaded.completed)
    }

    // ---- skip + resurface window ----------------------------------------------------------------

    @Test
    fun skip_stamps_time_and_gates_auto_open_within_the_window() {
        val storage = MemoryStorage()
        val skipAt = 1_000_000L
        val s = store(storage, now = skipAt)
        s.skip()
        assertEquals(skipAt, s.load().skippedAtMs)

        val steps = listOf("a", "b")
        // Within the window: quiet.
        assertFalse(shouldAutoOpenOnboarding(s.load(), steps, nowMs = skipAt + DEFAULT_RESURFACE_MS - 1))
        // At/after the window: may auto-open again.
        assertTrue(shouldAutoOpenOnboarding(s.load(), steps, nowMs = skipAt + DEFAULT_RESURFACE_MS))
        assertTrue(shouldAutoOpenOnboarding(s.load(), steps, nowMs = skipAt + DEFAULT_RESURFACE_MS + 1))
    }

    @Test
    fun fresh_state_auto_opens() {
        assertTrue(shouldAutoOpenOnboarding(OnboardingState(), listOf("a"), nowMs = 5_000L))
    }

    // ---- checklist done -------------------------------------------------------------------------

    @Test
    fun finished_checklist_suppresses_auto_open() {
        val steps = listOf("a", "b")
        val partial = OnboardingState(completed = listOf("a"))
        assertFalse(isChecklistDone(partial, steps))
        assertTrue(shouldAutoOpenOnboarding(partial, steps, nowMs = 5_000L))

        val done = OnboardingState(completed = listOf("a", "b"))
        assertTrue(isChecklistDone(done, steps))
        assertFalse(shouldAutoOpenOnboarding(done, steps, nowMs = 5_000L))

        // An empty step list is never "done" and does not gate auto-open (What's-new-only case).
        assertTrue(shouldAutoOpenOnboarding(OnboardingState(), emptyList(), nowMs = 5_000L))
    }

    // ---- suppressed -----------------------------------------------------------------------------

    @Test
    fun suppressed_stops_auto_open_and_every_tour() {
        val storage = MemoryStorage()
        val s = store(storage)
        s.setSuppressed(true)
        assertTrue(s.load().suppressed)
        assertFalse(shouldAutoOpenOnboarding(s.load(), listOf("a"), nowMs = 10_000_000L))
        assertFalse(shouldAutoStartTour(s.load(), "any-tour"))

        s.setSuppressed(false)
        assertTrue(shouldAutoStartTour(s.load(), "any-tour"))
    }

    // ---- tours ----------------------------------------------------------------------------------

    @Test
    fun finishTour_makes_the_tour_not_auto_start_and_records_the_outcome() {
        val storage = MemoryStorage()
        val s = store(storage)
        assertTrue(shouldAutoStartTour(s.load(), "rail-tour"))
        s.finishTour("rail-tour", TourOutcome.COMPLETED)
        assertFalse(shouldAutoStartTour(s.load(), "rail-tour"))
        assertEquals("completed", s.load().tours["rail-tour"])
        // A different tour is still eligible.
        assertTrue(shouldAutoStartTour(s.load(), "other-tour"))

        s.finishTour("other-tour", TourOutcome.SKIPPED)
        assertEquals("skipped", s.load().tours["other-tour"])
        assertFalse(shouldAutoStartTour(s.load(), "other-tour"))
    }

    // ---- markVersionSeen ------------------------------------------------------------------------

    @Test
    fun markVersionSeen_round_trips() {
        val storage = MemoryStorage()
        val s = store(storage)
        assertNull(s.load().lastSeenVersion)
        s.markVersionSeen("1.2.0")
        assertEquals("1.2.0", s.load().lastSeenVersion)
        assertEquals("1.2.0", store(storage).load().lastSeenVersion) // persisted
    }

    // ---- reset ----------------------------------------------------------------------------------

    @Test
    fun reset_clears_state_and_storage() {
        val storage = MemoryStorage()
        val s = store(storage)
        s.completeStep("a")
        s.setSuppressed(true)
        s.markVersionSeen("1.0.0")
        val cleared = s.reset()
        assertEquals(OnboardingState(), cleared)
        assertNull(storage.getItem(onboardingStorageKey("fudemoji", null)))
        // A fresh store over the same storage also sees the empty state.
        assertEquals(OnboardingState(), store(storage).load())
    }

    // ---- defensive parse ------------------------------------------------------------------------

    @Test
    fun parse_falls_back_to_empty_for_null_blank_and_garbage() {
        assertEquals(OnboardingState(), parseOnboardingState(null))
        assertEquals(OnboardingState(), parseOnboardingState(""))
        assertEquals(OnboardingState(), parseOnboardingState("   "))
        assertEquals(OnboardingState(), parseOnboardingState("not json at all {"))
        assertEquals(OnboardingState(), parseOnboardingState("[1,2,3]"))
    }

    @Test
    fun parse_tolerates_partial_and_unknown_fields() {
        // Only some fields present, plus an unknown key: known fields load, unknown is ignored, the rest
        // default. ignoreUnknownKeys keeps a forward/older record from throwing.
        val raw = """{"completed":["a"],"unknownFuture":true}"""
        val parsed = parseOnboardingState(raw)
        assertEquals(listOf("a"), parsed.completed)
        assertFalse(parsed.suppressed)
        assertNull(parsed.skippedAtMs)
        assertNull(parsed.lastSeenVersion)
        assertEquals(emptyMap(), parsed.tours)
    }

    @Test
    fun a_written_record_round_trips_through_parse() {
        val storage = MemoryStorage()
        val s = store(storage, now = 42L)
        s.completeStep("a")
        s.skip()
        s.finishTour("t", TourOutcome.COMPLETED)
        s.markVersionSeen("2.0.0")
        val onDisk = storage.getItem(onboardingStorageKey("fudemoji", null))
        val parsed = parseOnboardingState(onDisk)
        assertEquals(listOf("a"), parsed.completed)
        assertEquals(42L, parsed.skippedAtMs)
        assertEquals("completed", parsed.tours["t"])
        assertEquals("2.0.0", parsed.lastSeenVersion)
    }
}
